### PR TITLE
[`pygrep_hooks`] Check blanket ignores via file-level pragmas (`PGH004`)

### PR DESCRIPTION
**Pull request by ajesipow**
## Summary

Should resolve https://github.com/astral-sh/ruff/issues/11454.

This is my first PR to `ruff`, so I may have missed something.

If I understood the suggestion in the issue correctly, rule `PGH004` should be set to `Preview` again.

## Test Plan

Created two fixtures derived from the issue.

### CODE REVIEW

**Comment by JonathanPlasse** [comment]
The stable behavior should be unchanged, the added behavior should only be present in preview.
i.e. The ruff ecosystem results should not change for stable, only preview should have additional violations.
Otherwise, the users of stable will lose the PGH004 rule.

**Comment by ajesipow** [comment]
Thanks for the swift feedback.

I added a check for `PreviewMode::Enabled` and added tests for the preview case.

Would you like me to also update the docs for `BlanketNOQA`?

File: crates/ruff_linter/src/rules/pygrep_hooks/rules/blanket_noqa.rs [hunk 1]
```diff
@@ -85,7 +86,19 @@ pub(crate) fn blanket_noqa(
     diagnostics: &mut Vec<Diagnostic>,
     noqa_directives: &NoqaDirectives,
     locator: &Locator,
+    exemption: &Option<FileExemption>,
```
Comment (charliermarsh) [inline crates/ruff_linter/src/rules/pygrep_hooks/rules/blanket_noqa.rs:89]: One downside here is that this is only going to flag the _first_ `# ruff: noqa` in the file. I assume if you add multiple such comments, only the first is flagged as a diagnostic?

Comment (charliermarsh) [reply crates/ruff_linter/src/rules/pygrep_hooks/rules/blanket_noqa.rs:89]: We might need to do something like `NoqaDirectives`, whereby we parse all of these upfront, then use the parsed exemptions in `crates/ruff_linter/src/checkers/noqa.rs` (and here).

Comment (charliermarsh) [reply crates/ruff_linter/src/rules/pygrep_hooks/rules/blanket_noqa.rs:89]: @ajesipow - Do you want to give that a try?

Comment (ajesipow) [reply crates/ruff_linter/src/rules/pygrep_hooks/rules/blanket_noqa.rs:89]: Thanks for the ping @charliermarsh.
I haven't found the time yet after work this week, but today is a public holiday over here so I'll look into it today :)

File: crates/ruff_linter/src/noqa.rs [hunk 2]
```diff
@@ -275,26 +283,43 @@ pub(crate) fn rule_is_ignored(
...
-/// The file-level exemptions extracted from a given Python file.
+/// A summary of the file-level exemption as extracted from [`BlanketNoqaDirectives`].
 #[derive(Debug)]
-pub(crate) enum FileExemption {
+pub(crate) enum FileExemption<'a> {
     /// The file is exempt from all rules.
     All,
     /// The file is exempt from the given rules.
-    Codes(Vec<NoqaCode>),
+    Codes(Vec<&'a NoqaCode>),
```
Comment (ajesipow) [inline crates/ruff_linter/src/noqa.rs:292]: I left this as a `Vec`, but we could consider using a `HashSet` instead as we're checking if certain `NoqaCode`s are contained in a few places (one being the loop on line 54 in `checkers/noqa.rs`). 

It's probably fine though as I'd expect there will usually only be very few elements here and checking the small `Vec` is likely faster compared to the overhead of hashing every time (and building the `HashSet` in the first place).
We'd also have to derive `Hash` on `NoqaCode`.

File: crates/ruff_linter/src/rules/pygrep_hooks/rules/blanket_noqa.rs (hunk 1 above)
Comment (ajesipow) [reply crates/ruff_linter/src/rules/pygrep_hooks/rules/blanket_noqa.rs]: I just pushed the changes and adapted the tests again accordingly.

Looking forward to your feedback again.

**Comment by charliermarsh** [comment]
Thanks @ajesipow, I will take a look.

**Review by charliermarsh — APPROVED** [review]
Thank you!

**Pull request merged**